File: behavioral/specification.php
<?php
interface Specification {
    public function isSatisfiedBy(array $user): bool;
}

class AgeAbove18 implements Specification {
    public function isSatisfiedBy(array $user): bool {
        return $user['age'] > 18;
    }
}

class HasEmail implements Specification {
    public function isSatisfiedBy(array $user): bool {
        return !empty($user['email']);
    }
}

class AndSpec implements Specification {
    private Specification $a; private Specification $b;
    public function __construct(Specification $a, Specification $b){
        $this->a = $a; $this->b = $b;
    }
    public function isSatisfiedBy(array $user): bool {
        return $this->a->isSatisfiedBy($user) && $this->b->isSatisfiedBy($user);
    }
}

// استفاده
$user = ['age'=>20, 'email'=>'test@example.com'];
$spec = new AndSpec(new AgeAbove18(), new HasEmail());
echo $spec->isSatisfiedBy($user) ? "Ok" : "Rejected";

File: behavioral/specification.go
package main

import "fmt"

type User struct {
	Age   int
	Email string
}

type Specification interface {
	IsSatisfiedBy(User) bool
}

type AgeAbove18 struct{}

func (AgeAbove18) IsSatisfiedBy(u User) bool { return u.Age > 18 }

type HasEmail struct{}

func (HasEmail) IsSatisfiedBy(u User) bool { return u.Email != "" }

type AndSpec struct{ a, b Specification }

func (s AndSpec) IsSatisfiedBy(u User) bool {
	return s.a.IsSatisfiedBy(u) && s.b.IsSatisfiedBy(u)
}

func main() {
	user := User{Age: 20, Email: "test@example.com"}
	spec := AndSpec{AgeAbove18{}, HasEmail{}}
	if spec.IsSatisfiedBy(user) {
		fmt.Println("Ok")
	} else {
		fmt.Println("Rejected")
	}
}
